lls/main.c

```diff
@@ -316,6 +317,124 @@ lacp_timer_reset(struct lls_config *lls_conf, struct gatekeeper_if *iface)
 		lls_conf->lcore_id, lls_lacp_announce, iface);
 }
 
+static void
+fillup_lls_dump_entry(struct lls_dump_entry *dentry, struct lls_map *map,
+	uint32_t key_len)
+{
+	dentry->stale = map->stale;
+	dentry->port_id = map->port_id;
+	if (key_len == sizeof(struct in_addr)) {
+		dentry->ipaddr.proto = ETHER_TYPE_IPv4;
+		rte_memcpy(&dentry->ipaddr.ip.v4, map->ip_be,
+			sizeof(dentry->ipaddr.ip.v4));
+	} else if (likely(key_len == sizeof(struct in6_addr))) {
+		dentry->ipaddr.proto = ETHER_TYPE_IPv6;
+		rte_memcpy(&dentry->ipaddr.ip.v6, map->ip_be,
+			sizeof(dentry->ipaddr.ip.v6));
+	} else
+		dentry->ipaddr.proto = 0;
+	rte_memcpy(&dentry->ha, &map->ha, sizeof(dentry->ha));
+}
+
+#define CTYPE_STRUCT_LLS_DUMP_ENTRY_PTR "struct lls_dump_entry *"
+
+static void
+list_lls_cache(lua_State *l, struct lls_cache *cache)
+{
+	uint32_t next = 0;
+	const void *key;
+	void *data;
+	int32_t index;
+	void *cdata;
+	uint32_t correct_ctypeid_lls_dentry = luaL_get_ctypeid(l,
+		CTYPE_STRUCT_LLS_DUMP_ENTRY_PTR);
+
+	index = rte_hash_iterate(cache->hash, (void *)&key, &data, &next);
```

Comment:
Notice that this function is executed by the dynamic configuration block **concurrently** with the LLS block. So, what guarantees that this iteration over the cache table is free of race conditions? Adding a comment here in the code explaining that would be helpful.

Reply:
At the moment I actually don't think that there is any guarantee. A simple way to achieve this would be to enable the flag `RTE_HASH_EXTRA_FLAGS_RW_CONCURRENCY` for the cache (hash table) to support reader/writer concurrency between the Dynamic Config and LLS blocks. I think this is suitable for our purposes.

Reply:
I agree that `RTE_HASH_EXTRA_FLAGS_RW_CONCURRENCY` is enough here since the LLS block is not in the critical paths of Gatekeeper and Grantor servers.